xdsl/interactive/app.py

```diff
@@ -0,0 +1,139 @@
+
+Run `terminal -m xdsl.interactive.app:InputApp --def` to run in development mode. Please be sure to install `textual-dev` to run this command.
+"""
+
+from io import StringIO
+
+from rich.style import Style
+from textual import events, on
+from textual.app import App, ComposeResult
+from textual.containers import Horizontal, Vertical
+from textual.reactive import reactive
+from textual.widgets import Footer, TextArea
+from textual.widgets.text_area import TextAreaTheme
+
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import MLContext
+from xdsl.parser import Parser
+from xdsl.printer import Printer
+from xdsl.tools.command_line_tool import get_all_dialects
+
+
+class OutputTextArea(TextArea):
+    """Used to prevent users from being able to change/alter the Output TextArea"""
+
+    async def _on_key(self, event: events.Key) -> None:
+        event.prevent_default()
+
+
+class InputApp(App[None]):
+    """
+    Interactive application for constructing compiler pipelines.
+    """
+
+    CSS_PATH = "app.tcss"
+
+    BINDINGS = [
+        ("d", "toggle_dark", "Toggle dark mode"),
+        ("q", "quit_app", "Quit"),
+    ]
+
+    # defines a theme for the Input/Output TextArea's
+    _DEFAULT_THEME = TextAreaTheme(
+        name="my_theme_design",
+        base_style=Style(bgcolor="white"),
+        syntax_styles={
+            "string": Style(color="red"),
+            "comment": Style(color="magenta"),
+        },
+    )
+
+    current_module: reactive[ModuleOp | Exception | None] = reactive(None)
+    """
+    Reactive variable used to save the current state of the modified Input TextArea (i.e. is the Output TextArea)
+    """
+
+    input_text_area = TextArea(id="input")
+    output_text_area = OutputTextArea(id="output")
+
+    def compose(self) -> ComposeResult:
+        """
+        Creates the required widgets, events, etc.
+        Get the list of xDSL passes, add them to an array in "Selection" format (so it can be added to a Selection List)
+        and sort the list in alphabetical order.
+        """
+
+        with Horizontal(id="input_output"):
+            with Vertical(id="input_container"):
+                yield self.input_text_area
+            with Vertical(id="output_container"):
+                yield self.output_text_area
+        yield Footer()
+
+    @on(TextArea.Changed, "#input")
+    def update_current_module(self) -> None:
+        """
+        Function called when the Input TextArea is cahnged. This function parses the Input IR and updates
+        the current_module reactive variable.
+        """
+        input_text = self.input_text_area.text
+        try:
+            ctx = MLContext(True)
+            for dialect in get_all_dialects():
+                ctx.load_dialect(dialect)
+            parser = Parser(ctx, input_text)
+            module = parser.parse_module()
+            self.current_module = module
+        except Exception as e:
+            self.current_module = e
```

Comment:
I think this function should be in an individual class that implements the logic behind the tool.